core/src/main/java/org/apache/calcite/rel/metadata/FunctionalDependencySet.java

```diff
@@ -0,0 +1,369 @@
+        }
+        missing = missing - 1;
+        fdMissingCount.put(fd, missing);
+        if (missing == 0) {
+          for (int dep : fd.getDependents()) {
+            if (closureSet.add(dep)) {
+              queue.add(dep);
+            }
+          }
+        }
+      }
+    }
+
+    return ImmutableBitSet.of(closureSet);
+  }
+
+  /**
+   * Check if X determined Y is implied by this FD set.
+   */
+  public boolean implies(ImmutableBitSet determinants, ImmutableBitSet dependents) {
+    return closure(determinants).contains(dependents);
+  }
+
+  /**
+   * Check if a single column is functionally determined by another column.
+   */
+  public boolean determines(int determinant, int dependent) {
+    return closure(ImmutableBitSet.of(determinant)).get(dependent);
+  }
+
+  /**
+   * Compute the minimal cover of this functional dependency set.
+   * Returns an equivalent set with minimal dependencies.
+   */
+  public FunctionalDependencySet minimalCover() {
+    // Split multi-attribute right sides into single attributes
+    Set<FunctionalDependency> splitFDs = new HashSet<>();
+    for (FunctionalDependency fd : fdSet) {
+      splitFDs.addAll(fd.split());
+    }
+    splitFDs.removeIf(FunctionalDependency::isTrivial);
+
+    // Remove redundant attributes from left sides
+    Set<FunctionalDependency> reducedFDs = new HashSet<>();
+    for (FunctionalDependency fd : splitFDs) {
+      FunctionalDependencySet tempSet = new FunctionalDependencySet(splitFDs);
+      tempSet.removeFD(fd);
+      reducedFDs.add(reduceLeft(fd, tempSet));
+    }
+
+    // Remove redundant functional dependencies
+    reducedFDs.removeIf(fd -> {
+      FunctionalDependencySet remainingFDs = new FunctionalDependencySet(reducedFDs);
+      remainingFDs.removeFD(fd);
+      return remainingFDs.implies(fd.getDeterminants(), fd.getDependents());
+    });
+
+    return new FunctionalDependencySet(reducedFDs);
+  }
+
+  /**
+   * Reduce left side by removing redundant columns from determinants.
+   */
+  private static FunctionalDependency reduceLeft(FunctionalDependency fd,
+      FunctionalDependencySet fdSet) {
+    ImmutableBitSet determinants = fd.getDeterminants();
+    ImmutableBitSet dependents = fd.getDependents();
+
+    // Try removing each attribute to find minimal determinant set
+    for (int attr : fd.getDeterminants()) {
+      ImmutableBitSet reduced = determinants.clear(attr);
+      if (fdSet.closure(reduced).contains(dependents)) {
+        determinants = reduced;
+      }
+    }
+    return new FunctionalDependency(determinants, dependents);
+  }
+
+  /**
+   * Check if this FD set is equivalent to another FD set.
+   * Two FD sets are equivalent if they have the same closure for any attribute set.
+   */
+  public boolean equalTo(FunctionalDependencySet other) {
+    // Check if every FD in this set is implied by the other set
+    for (FunctionalDependency fd : fdSet) {
+      if (!other.implies(fd.getDeterminants(), fd.getDependents())) {
```

Comment:
implies calls closure, so this test looks quite expensive in the worst case.

Reply:
There are already some test cases for the closure. Do we need to add additional test cases for this? Essentially, it's still a matter of the closure's execution efficiency.